netket/utils/semigroup.py

```diff
@@ -187,3 +186,145 @@ def __repr__(self):
         else:
             elems = map(repr, self.elems)
         return type(self).__name__ + "(\n  {}\n)".format(",\n  ".join(elems))
+
+
+@dataclass(frozen=True)
+class PermutationGroup(SemiGroup):
+    """
+    Collection of permutation operations acting on sequences of length :code:`degree`.
+    Note that the group elements do not need to be of type :ref:`netket.utils.semigroup.Permutation`,
+    only act as such on a sequence when called.
+
+    Note that this class can contain elements that are distinct as objects (e.g., :code:`Identity()` and
+    :code:`Translation((0,))`) but have identical action. Those can be removed by calling :code:`remove_duplicates`.
+    """
+
+    degree: int
+    """Number of elements the permutations act on"""
+
+    def __post_init__(self):
+        super().__post_init__()
+        myhash = hash((super().__hash__(), hash(self.degree)))
+        object.__setattr__(self, "_PermutationGroup__hash", myhash)
+
+        object.__setattr__(self, "_inverse", None)
+        object.__setattr__(self, "_product_table", None)
+
+    def __matmul__(self, other):
+        if isinstance(other, PermutationGroup) and self.degree != other.degree:
+            raise ValueError(
+                "Incompatible groups (`PermutationGroup`s of different degree)"
+            )
+
+        return PermutationGroup(super().__matmul__(other).elems, self.degree)
+
+    def to_array(self):
+        """
+        Convert the abstract group operations to an array of permutation indicies,
+        such that (for :code:`G = self`)::
+            V = np.arange(G.degree)
+            assert np.all(G(V) == V[..., G.to_array()])
+        """
```

Comment:
Could we also state this in maths (rather than coding) language? I spent an awful lot of time understanding that a `Permutation` lists the preimages of each site, and I'm not sure whether a group element here corresponds to a row or a column...

Reply:
Maybe like this?
```suggestion
        """
        Convert the group to an array of permutation indices, such that the `i`-th row contains the
        indices corresponding to the `i`-th group element and (for :code:`G = self`)::
            V = np.arange(G.degree)
            assert np.all(G(V) == V[..., G.to_array()])
        """
```

Reply:
```suggestion
        """
        Convert the abstract group operations to an array of permutation indices, 
        such that the `i`-th row contains the indices corresponding to the `i`-th group 
        element (i.e., `self.to_array()[i,j]` gives gᵢ⁻¹(j)), and (for :code:`G = self`)::
            V = np.arange(G.degree)
            assert np.all(G(V) == V[..., G.to_array()])
        """
```
It's best to document what each row means, too!